src/ResultWriter/EnergyOutput.cpp

```diff
@@ -98,57 +167,243 @@ void seissol::writer::printEnergies(GlobalData const* global,
+#ifdef USE_ELASTIC
+    real volume = MeshTools::volume(elements[elementId], vertices);
+    CellMaterialData& material = ltsLut->lookup(lts->material, elementId);
+    auto& cellInformation = ltsLut->lookup(lts->cellInformation, elementId);
+    auto& faceDisplacements = ltsLut->lookup(lts->faceDisplacements, elementId);
+
+    constexpr auto quadPolyDegree = CONVERGENCE_ORDER + 1;
+    constexpr auto numQuadraturePointsTet = quadPolyDegree * quadPolyDegree * quadPolyDegree;
+
+    double quadraturePointsTet[numQuadraturePointsTet][3];
+    double quadratureWeightsTet[numQuadraturePointsTet];
+    seissol::quadrature::TetrahedronQuadrature(
+        quadraturePointsTet, quadratureWeightsTet, quadPolyDegree);
+
+    constexpr auto numQuadraturePointsTri = quadPolyDegree * quadPolyDegree;
+    double quadraturePointsTri[numQuadraturePointsTri][2];
+    double quadratureWeightsTri[numQuadraturePointsTri];
+    seissol::quadrature::TriangleQuadrature(
+        quadraturePointsTri, quadratureWeightsTri, quadPolyDegree);
+
+    // Needed to weight the integral.
+    const auto jacobiDet = 6 * volume;
+
+    alignas(ALIGNMENT) real numericalSolutionData[tensor::dofsQP::size()];
+    auto numericalSolution = init::dofsQP::view::create(numericalSolutionData);
+    // Evaluate numerical solution at quad. nodes
+    kernel::evalAtQP krnl;
+    krnl.evalAtQP = global->evalAtQPMatrix;
+    krnl.dofsQP = numericalSolutionData;
+    krnl.Q = ltsLut->lookup(lts->dofs, elementId);
+    krnl.execute();
+
+#ifdef MULTIPLE_SIMULATIONS
+    auto numSub = numericalSolution.subtensor(sim, yateto::slice<>(), yateto::slice<>());
 #else
-  rank = 0;
-  totalFrictionalWorkGlobal = totalFrictionalWorkLocal;
-  staticFrictionalWorkGlobal = staticFrictionalWorkLocal;
-  plasticMomentGlobal = plasticMomentLocal;
+    auto numSub = numericalSolution;
+#endif
+    for (size_t qp = 0; qp < numQuadraturePointsTet; ++qp) {
+      constexpr int uIdx = 6;
+      const auto curWeight = jacobiDet * quadratureWeightsTet[qp];
+      const auto rho = material.local.rho;
+
+      const auto u = numSub(qp, uIdx + 0);
+      const auto v = numSub(qp, uIdx + 1);
+      const auto w = numSub(qp, uIdx + 2);
+      const double curKineticEnergy = 0.5 * rho * (u * u + v * v + w * w);
+
+      if (std::abs(material.local.mu) < 10e-14) {
+        // Acoustic
+        constexpr int pIdx = 0;
+        const auto K = material.local.lambda;
+        const auto p = numSub(qp, pIdx);
+
+        const double curAcousticEnergy = (p * p) / ( 2 * K);
+        totalAcousticEnergyLocal += curWeight * curAcousticEnergy;
+        totalAcousticKineticEnergyLocal += curWeight * curKineticEnergy;
+      } else {
+        // Elastic
+        totalElasticKineticEnergyLocal += curWeight * curKineticEnergy;
+        auto getStressIndex = [](int i, int j) {
+          auto lookup = std::array<std::array<int, 3>, 3>{{{0, 3, 5}, {3, 1, 4}, {5, 4, 2}}};
```

Comment:
Would it be worth initializing this array outside the function and the loops and passing it as a reference?

Reply:
The compiler should recognize that it is constant. I'll mark it static const to enforce this